client/network/src/protocol.rs

```diff
@@ -179,6 +179,10 @@ pub struct Protocol<B: BlockT, Client> {
 	/// List of nodes for which we perform additional logging because they are important for the
 	/// user.
 	important_peers: HashSet<PeerId>,
+	/// List of nodes that should never occupy peer slots.
+	default_peers_set_no_slot_peers: HashSet<PeerId>,
+	/// Actual list of connected no-slot nodes.
+	default_peers_set_no_slot_connected_peers: HashSet<PeerId>,
```

Comment:
Looks like this could be a simple counter, to be Incremented/decremented when the peer is in the reserved set on connect/disconnect.

Reply:
This is how I implemented it originally in [2f49625](https://github.com/paritytech/substrate/pull/11909/commits/2f49625ff6c19cb6a3fea2d493c49f059f51befc). Later I replaced the counter with a more fool-proof set in [0336aaf](https://github.com/paritytech/substrate/pull/11909/commits/0336aaf995b44c69e2e05a605059de5efc6c2d54) (the set will work even if we modify the list of reserved nodes, while the counter clearly won't). May be this is unnecessary.

Reply:
I agree that this is more fool proof. Let's leave it.